### ISSUE TITLE
Ordering products in the admin times out when having too many products

### ISSUE THREAD

**Issue by Tofandel** [woocommerce/woocommerce#30995]
**Prerequisites (mark completed items with an \[x\]):**

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**<br>When having large numbers of products (4000+) using the sort feature in the product list, the request will take more than 5 minutes to complete because it fetches the menu order of all products and updates the order of all 4000 products one by one in one request and so times out in many setup as well make the feature clearly unusable from a UX standpoint

The issue lies within [this piece of code](<https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-ajax.php#L1868>) as you can see

```php
$menu_orders = wp_list_pluck( $wpdb->get_results( "SELECT ID, menu_order FROM {$wpdb->posts} WHERE post_type = 'product' ORDER BY menu_order ASC, post_title ASC" ), 'menu_order', 'ID' );
$index       = 0;

foreach ( $menu_orders as $id => $menu_order ) {
	$id = absint( $id );

	if ( $sorting_id === $id ) {
		continue;
	}
	if ( $nextid === $id ) {
		$index ++;
	}
	$index ++;
	$menu_orders[ $id ] = $index;
	$wpdb->update( $wpdb->posts, array( 'menu_order' => $index ), array( 'ID' => $id ) );

	/**
		* When a single product has gotten it's ordering updated.
		* $id The product ID
		* $index The new menu order
	*/
	do_action( 'woocommerce_after_single_product_ordering', $id, $index );
}
```

A different approach should be used to sort products efficiently that only needs to fetch and update a limited number of products (3) by using this method [https://stackoverflow.com/a/25432597/2977175](<https://stackoverflow.com/a/25432597/2977175>)

**Expected behavior**<br>The sort to complete within less than 10 seconds in all setup and not just on sites with less than 100 products

**Actual behavior**

```
POST /wp-admin/admin-ajax.php 504 Gateway Timeout 60071ms
action | "woocommerce_product_ordering"
id | "165506"
previd | "153572"
nextid | "163317"
```

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**<br>Steps to reproduce the bug:

1. Insert more than 5000 products
2. Go to 'Products > Sorting'
3. Move up or down one product

**Isolating the problem (mark completed items with an \[x\]):**

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](<https://woocommerce.com/storefront/>).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by Tofandel** [comment]
I have been experimenting with a new function that seems to work really well

```php
/**
 * Ajax request handling for product ordering.
 *
 * Based on Simple Page Ordering by 10up (https://wordpress.org/plugins/simple-page-ordering/).
 */
public static function product_ordering()
{
    global $wpdb;

    // phpcs:disable WordPress.Security.NonceVerification.Missing
    if (!current_user_can('edit_products') || empty($_POST['id'])) {
        wp_die(-1);
    }

    $sorting_id = absint($_POST['id']);
    $previd = absint(isset($_POST['previd']) ? $_POST['previd'] : 0);
    $nextid = absint(isset($_POST['nextid']) ? $_POST['nextid'] : 0);
    $menu_orders = wp_list_pluck($wpdb->get_results("SELECT ID, menu_order FROM {$wpdb->posts} WHERE post_type = 'product' AND ID IN ($previd, $nextid, $sorting_id)"), 'menu_order', 'ID');

    $normalize = function ($id) use ($menu_orders) {
        return $id && $menu_orders[$id] ? $menu_orders[$id] : 0;
    };

    if ($previd && !$menu_orders[$previd]) {
        $menu_orders[$previd] = (int)round(PHP_INT_MAX / 2);
        $wpdb->update($wpdb->posts, array('menu_order' => $menu_orders[$previd]), array('ID' => $previd));
        do_action('woocommerce_after_single_product_ordering', $previd, $menu_orders[$previd]);
    }
    if ($nextid && !$menu_orders[$nextid]) {
        $menu_orders[$nextid] = (int)round(PHP_INT_MAX / 2);
        $wpdb->update($wpdb->posts, array('menu_order' => $menu_orders[$nextid]), array('ID' => $previd));
        do_action('woocommerce_after_single_product_ordering', $nextid, $menu_orders[$nextid]);
    }

    $menu_orders[$sorting_id] = (int)round($normalize($nextid) / 2 + $normalize($previd) / 2);

    /**
     * When a single product has gotten it's ordering updated.
     * $id The product ID
     * $index The new menu order
     */
    $wpdb->update($wpdb->posts, array('menu_order' => $menu_orders[$sorting_id]), array('ID' => $sorting_id));
    do_action('woocommerce_after_single_product_ordering', $sorting_id, $menu_orders[$sorting_id]);


    WC_Post_Data::delete_product_query_transients();

    do_action('woocommerce_after_product_ordering', $sorting_id, $menu_orders);
    wp_send_json($menu_orders);
    // phpcs:enable
}
```

With the condition that the menu_order column needs to be changed to a BIG INT (20) instead of the wp default INT(11) (it could be adapted to work with INT (11) by changing the PHP_INT_MAX to 2 ^ 31

And that the menu order also need a migration to multiply all the orders by `PHP_INT_MAX / number of products` to enlargen the range and that a newly created product be set with an order of `order of previous product / 2` or `PHP_INT_MAX / 2`

**Comment by juliaamosova** [comment]
Hi @Tofandel,

Thank you for opening the issue! It requires further feedback from the WooCommerce Core team. I am adding the `needs developer feedback` label to this issue so that the Core team could take a look. 

Please note it may take a few days for them to get to this issue. Thank you for your patience.

**Comment by vedanshujain** [comment]
Hi @Tofandel , thanks for the report and suggestions. We would definitely want to improve this, since you already have a suggestion, would you like to send us a PR so that we can take a detailed look?

**Comment by theo-gk** [comment]
Hi, is there any update on this? Still a major issue in store with thousands of products.

**Comment by kristos80** [comment]
This is a major issue and doesn't seem to get the attention needed. 

```php
		if($wpdb->update($wpdb->posts, ['menu_order' => $menu_orders[$sorting_id]], ['ID' => $sorting_id])) {
			// We only need to clean the cache if the menu order was actually modified.
			clean_post_cache($sorting_id);
		}
```

The code is optimal when you are setting the order while being in the full list of products. If you are using filters of any kind, then a lot more products will be needed to get a new `menu_order` value, which will trigger the `clean_post_cache` that affects performance.

The sorting indexes shouldn't be part of the `post` record so that it doesn't need to update the whole record every time it gets recalculated.

**Comment by kristos80** [comment]
> I have been experimenting with a new function that seems to work really well
> 
> /**
>  * Ajax request handling for product ordering.
>  *
>  * Based on Simple Page Ordering by 10up (https://wordpress.org/plugins/simple-page-ordering/).
>  */
> public static function product_ordering()
> {
>     global $wpdb;
> 
>     // phpcs:disable WordPress.Security.NonceVerification.Missing
>     if (!current_user_can('edit_products') || empty($_POST['id'])) {
>         wp_die(-1);
>     }
> 
>     $sorting_id = absint($_POST['id']);
>     $previd = absint(isset($_POST['previd']) ? $_POST['previd'] : 0);
>     $nextid = absint(isset($_POST['nextid']) ? $_POST['nextid'] : 0);
>     $menu_orders = wp_list_pluck($wpdb->get_results("SELECT ID, menu_order FROM {$wpdb->posts} WHERE post_type = 'product' AND ID IN ($previd, $nextid, $sorting_id)"), 'menu_order', 'ID');
> 
>     $normalize = function ($id) use ($menu_orders) {
>         return $id && $menu_orders[$id] ? $menu_orders[$id] : 0;
>     };
> 
>     if ($previd && !$menu_orders[$previd]) {
>         $menu_orders[$previd] = (int)round(PHP_INT_MAX / 2);
>         $wpdb->update($wpdb->posts, array('menu_order' => $menu_orders[$previd]), array('ID' => $previd));
>         do_action('woocommerce_after_single_product_ordering', $previd, $menu_orders[$previd]);
>     }
>     if ($nextid && !$menu_orders[$nextid]) {
>         $menu_orders[$nextid] = (int)round(PHP_INT_MAX / 2);
>         $wpdb->update($wpdb->posts, array('menu_order' => $menu_orders[$nextid]), array('ID' => $previd));
>         do_action('woocommerce_after_single_product_ordering', $nextid, $menu_orders[$nextid]);
>     }
> 
>     $menu_orders[$sorting_id] = (int)round($normalize($nextid) / 2 + $normalize($previd) / 2);
> 
>     /**
>      * When a single product has gotten it's ordering updated.
>      * $id The product ID
>      * $index The new menu order
>      */
>     $wpdb->update($wpdb->posts, array('menu_order' => $menu_orders[$sorting_id]), array('ID' => $sorting_id));
>     do_action('woocommerce_after_single_product_ordering', $sorting_id, $menu_orders[$sorting_id]);
> 
> 
>     WC_Post_Data::delete_product_query_transients();
> 
>     do_action('woocommerce_after_product_ordering', $sorting_id, $menu_orders);
>     wp_send_json($menu_orders);
>     // phpcs:enable
> }
> With the condition that the menu_order column needs to be changed to a BIG INT (20) instead of the wp default INT(11) (it could be adapted to work with INT (11) by changing the PHP_INT_MAX to 2 ^ 31
> 
> And that the menu order also need a migration to multiply all the orders by `PHP_INT_MAX / number of products` to enlargen the range and that a newly created product be set with an order of `order of previous product / 2` or `PHP_INT_MAX / 2`

Based on this solution, I adapted to something that changes only the `sorting_id` based on the `menu_order` of `prev_id` and `next_id` according to this logic:

- If only `prev_id` exists then the new `menu_order` of `sorting_id` will be `prev_id`.`menu_order` + 1;
- If only `next_id` exists then the new `menu_order` of `sorting_id` will be `next_id`.`menu_order` - 1, but never smaller than 0;
- If both `prev_id` and `next_id` exists then If `sorting_id`.`menu_order` < `prev_id`.`menu_order`, then the new `menu_order` will be between `prev_id`.`menu_order` and `next_id`.`menu_order`, but always closer to `next_id` and vice versa.

```php

public static function product_ordering() {
	global $wpdb;

	// phpcs:disable WordPress.Security.NonceVerification.Missing
	if(!current_user_can('edit_products') || empty($_POST['id'])) {
		wp_die(-1);
	}

	$sorting_id = absint($_POST['id']);
	$prev_id = absint($_POST['previd'] ?? 0);
	$next_id = absint($_POST['nextid'] ?? 0);

	$menu_orders =
		wp_list_pluck($wpdb->get_results("SELECT ID, menu_order FROM $wpdb->posts WHERE post_type = 'product' AND ID IN ($prev_id, $next_id, $sorting_id)"),
			'menu_order',
			'ID');

	$prev_order = $prev_id !== 0 ? ($menu_orders[$prev_id] ?? NULL) : NULL;
	$next_order = $next_id !== 0 ? ($menu_orders[$next_id] ?? NULL) : NULL;
	$current_order = $menu_orders[$sorting_id] ?? NULL;

	if(!$current_order) {
		return $menu_orders ?? [];
	}

	if($prev_order !== NULL && $next_order !== NULL) {
		if($current_order < $prev_order) {
			$diff = (int) ceil(($next_order - $prev_order) / 2);
			$new_order = $prev_order + $diff;
			if($new_order <= $prev_order) {
				$new_order = $next_order;
			}
		} else {
			$diff = (int) floor(($next_order - $prev_order) / 2);
			$new_order = $next_order - $diff;
			if($new_order >= $next_order) {
				$new_order = $prev_order;
			}
		}
	} elseif($prev_order !== NULL) {
		$new_order = $prev_order + 1;
	} elseif($next_order !== NULL) {
		$new_order = $next_order - 1;
		if($new_order <= 0) {
			$new_order = 0;
		}
	} else {
		$new_order = $current_order;
	}

	$new_menu_orders = [];
	if($prev_id !== 0) {
		$new_menu_orders[$prev_id] = $menu_orders[$prev_id];
	}

	$new_menu_orders[$sorting_id] = $new_order;

	if($next_id !== 0) {
		$new_menu_orders[$next_id] = $menu_orders[$next_id];
	}

	if($wpdb->update($wpdb->posts, ['menu_order' => $new_order], ['ID' => $sorting_id])) {
		clean_post_cache($sorting_id);
	}

	WC_Post_Data::delete_product_query_transients();

	do_action('woocommerce_after_product_ordering', $sorting_id, $new_menu_orders);
	wp_send_json($new_menu_orders);
	// phpcs:enable
}
```
It's not tested and it's only a suggestion towards a specific direction

**Comment by Tofandel** [comment]
Another easier solution without a migration could be to move the

```
	$wpdb->update( $wpdb->posts, array( 'menu_order' => $index ), array( 'ID' => $id ) );
```

Outside of the loop, we need to compile an update command that updates all the rows at once or runs all the queries in a single transaction (transactions are not supported on myIsam which wp supports, so only the first solution might be viable)

Like this Mysql only needs to lock the table and rewrite indexes once for the whole transaction

This would likely solve the timeout issue

**Comment by kristos80** [comment]
In our case it's not a problem the write operations, but cache invalidation `clean_post_cache($sorting_id)`; 

But it might be related as the write operations might affect the overall performance.

**Comment by Tofandel** [comment]
At the time I opened the issue, clean_post_cache was not called there  https://github.com/woocommerce/woocommerce/commit/7021d442bdc97e5eaa641d0241d39de195274cd0

So I guess performance is even worse now, the cache invalidation should likely be scheduled and be done in a cron, not there. But yeah invalidating the cache of basically all the products, when you reorder just one product is a big design issue
